graph/patterns/upsert.md

```diff
@@ -0,0 +1,209 @@
+
+## Problem
+
+Infrastructure as code (IaC) defines system resources and topologies in a declarative manner that allows teams to manage those resources as they would code.
+Practicing IaC helps teams deploy system resources in a reliable, repeatable, and controlled way. 
+IaC also helps automate deployment and reduces the risk of human error, especially for complex large environments. 
+Customers want to adopt IaC practices for many of the resources managed through Microsoft Graph.
+
+Most resources' creation operations in Microsoft Graph are not idempotent in nature.
+As a consequence, API consumers that want to offer IaC solutions, must create compensation layers that can mimic idempotent behavior. 
+For example, when creating a resource, the compensation layer must check whether the resource first exists, before trying to create or update the resource.
+
+Additionally, IaC code scripts or templates usually employ client-provided names (or keys) to track resources in a predictable manner, whereas [Microsoft Graph guidelines](../GuidelinesGraph.md#behavior-modeling) suggests use of `POST` to create new entities with service-generated keys.
+
+## Solution
+
+The solution is to use an `Upsert` pattern, to solve for the non-idempotent creation and client-provided naming problems.
+
+* For IaC scenarios, resources must use `Upsert` semantics with an [alternate key](./alternate-key.md):
+  * Use `PATCH` with a client-provided alternate key.
+  * For a non-existent resource (specified by the alternate key) the service must handle this as a "create". As part of creation, the service must still generate the primary key value.
+  * For an existing resource (specified by the alternate key) the service must handle this as an "update.
+  * Any new alternate key, used for IaC scenarios, should be called `uniqueName`, if there isn't already an existing property that could be used as an alternate key.
+* NOTE: the service must also support `GET` using the alternate key pattern.
+* For consistent CRUD Microsoft Graph behaviors, all resources, **including** resources used in IaC scenarios, should use `POST` and a service-generated primary key, per existing guidelines, and support `GET`, `PATCH` and `DELETE` using the primary key.
+* If a service does not support `Upsert`, then a `PATCH` call against a non-existent resource must result in an HTTP "409 conflict" error.
+
+This solution allows for existing resources that follow Microsoft Graph conventions for CRUD operations to add `Upsert` without impacting existing apps or functionality.
+
+Ideally, all new entity types should support an `Upsert` mechanism, especially where they support control-plane APIs, or are used in admin style or IaC scenarios.
+
+## When to use this pattern
+
+This pattern should be adopted for resources that are managed through infrastructure as code or desired state configuration.
+
+## Issues and considerations
```

Comment:
>Issues and considerations [](http://example.com/codeflow?start=3&length=25)

Do we want to mention forcing update behavior with If-Match="*" (fails if already exists)?  What about forcing create behavior with If-None-Match="*"?

I think the former is useful, the later probably not terribly useful.

Reply:
Is supporting these headers required? Not sure either is useful, given the current behaviour.

For IaC scenarios, the point is that the template doesn't know whether the resource is created or not, so the if-Match and If-None_Match headers aren't particularly useful.

Reply:
This is not needed for IaC scenarios, and fail if exists/forcing create are not very interesting, but if the service does support upsert, and the client wants to be able to update a resource only if it already exists (not create if it doesn't) then the If-Match=* is useful.

Reply:
Updated